File: clients/go/logz.go
package logz

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"strings"
	"sync"
	"time"
)

var (
	serverAddr = getEnvOrDefault("LOGZ_ADDR", "127.0.0.1:12345")
	hostname   string
	logChan    = make(chan logEntry, 1000)
	once       sync.Once
)

type logEntry struct {
	srcLine string
	message string
}

func getEnvOrDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func init() {
	hostname, _ = os.Hostname()
	go backgroundSender()
}

// Log sends a message to the logz server
func Log(src, message string) {
	srcLine := fmt.Sprintf("%s %s %s", time.Now().Format("15:04:05"), hostname, src)
	select {
	case logChan <- logEntry{srcLine: srcLine, message: message}:
	default:
		// Channel full, drop newest (don't block)
	}
}

func backgroundSender() {
	var conn net.Conn
	var writer *bufio.Writer

	for {
		// Connect if not connected
		if conn == nil {
			var err error
			conn, err = net.DialTimeout("tcp", serverAddr, 5*time.Second)
			if err != nil {
				time.Sleep(1 * time.Second)
				continue
			}
			writer = bufio.NewWriter(conn)
		}

		// Read from channel
		entry := <-logChan

		err := sendMessage(writer, entry.srcLine, entry.message)
		if err != nil {
			// Connection lost, close and retry
			conn.Close()
			conn = nil
			writer = nil
			// Re-queue the message (non-blocking)
			select {
			case logChan <- entry:
			default:
			}
			time.Sleep(1 * time.Second)
		}
	}
}

func sendMessage(writer *bufio.Writer, srcLine, message string) error {
	// Write source line
	if _, err := writer.WriteString(srcLine + "\n"); err != nil {
		return err
	}

	// Write message lines, escaping __END__
	normalized := strings.ReplaceAll(strings.ReplaceAll(message, "\r\n", "\n"), "\r", "\n")
	lines := strings.Split(normalized, "\n")
	for _, line := range lines {
		if line == "__END__" {
			line = "__END____END__"
		}
		if _, err := writer.WriteString(line + "\n"); err != nil {
			return err
		}
	}

	// Write terminator and flush
	if _, err := writer.WriteString("__END__\n"); err != nil {
		return err
	}
	return writer.Flush()
}

File: clients/typescript/logz.ts
import * as net from 'net'
import * as os from 'os'

const SERVER_HOST = process.env.LOGZ_HOST ?? '127.0.0.1'
const SERVER_PORT = parseInt(process.env.LOGZ_PORT ?? '12345', 10)
const HOSTNAME = os.hostname()

type QueueEntry = { srcLine: string; message: string }

const queue: QueueEntry[] = []
let socket: net.Socket | null = null
let connecting = false
let draining = false

function ensureConnection() {
  if (socket || connecting) return
  connecting = true

  const s = new net.Socket()
  s.setEncoding('utf-8')

  s.connect(SERVER_PORT, SERVER_HOST, () => {
    socket = s
    connecting = false
    drain()
  })

  s.on('error', () => {
    socket = null
    connecting = false
    s.destroy()
    // Retry after delay
    setTimeout(ensureConnection, 1000)
  })

  s.on('close', () => {
    socket = null
    connecting = false
  })
}

function drain() {
  if (draining || !socket) return
  draining = true

  while (queue.length > 0 && socket) {
    const entry = queue.shift()!
    const escaped = entry.message
      .replace(/\r\n/g, '\n')
      .split('\n')
      .map((line) => (line === '__END__' ? '__END____END__' : line))
      .join('\n')

    try {
      socket.write(`${entry.srcLine}\n${escaped}\n__END__\n`)
    } catch {
      // Re-queue on write failure
      queue.unshift(entry)
      socket?.destroy()
      socket = null
      break
    }
  }

  draining = false
}

export function logz(src: string, message: string) {
  const ts = new Date().toLocaleTimeString('en-GB', { hour12: false })
  const srcLine = `${ts} ${HOSTNAME} ${src}`

  // Cap queue size
  if (queue.length < 1000) {
    queue.push({ srcLine, message })
  }

  ensureConnection()
  drain()
}
